Add Quartz Scheduler support

vpavic opened the pull request:
Spring Framework has had Quartz Scheduler support for a very long time - it would be nice to make use of it in Spring Boot.

This PR adds:
- auto-configuration for `SchedulerFactoryBean`
- `spring-boot-starter-quartz`
- `spring-boot-sample-quartz`
- relevant documentation updates

Please review, comments are welcome.
If this PR is accepted I'd also like to add Actuator support in a separate PR.

I've signed the CLA.

snicoll commented:
See also #3919 (scheduling, async processing theme).

vpavic commented:
Scheduling/async theme sounds good.

I was hoping this PR could sneak into the ```1.3.0.RELEASE```, but looking at the [release schedule](https://github.com/spring-projects/spring-boot/milestones) I guess it's too late for that now.

snicoll commented:
Yes, we are post RC1 now.

philwebb commented:
Since there's quite a lot of code in the pull-request we'd like to wait for a while and see how much demand there is for quartz. We'll mark this one as `waiting-for-votes` for now and monitor for +1's.

gnuphie commented:
+1 we use quartz all the time

iturcino commented:
+1 , we also need quartz support

beegor commented:
+1, this would be very useful

sschedl commented:
+1, would be very useful

vpugar commented:
+1 would be very useful

thackel commented:
+1, I don't know any other free solution with the feature set and maturity like quartz.

davidkiss commented:
+1, adding first-class support to quartz would be very helpful when working with this framework

vpavic commented:
Please reconsider adding this to one of ```1.4.0``` milestone releases.
User interest obviously exists, and Spring Boot doesn't provide much auto-configuration support in the area of advanced job scheduling.

Regarding the lot of code argument, IMO implementation is pretty straightforward and it is based on Quartz support in the core Spring Framework, test coverage is good, sample app and documentation are also in the mix so it shouldn't be a risky addition.

If this is added to the upcoming milestones, I can rebase onto current master and update the PR to fix the conflicts.

snicoll commented:
> Regarding the lot of code argument, IMO implementation is pretty straightforward and it is based on Quartz support in the core Spring Framework

Well IMO, that's exactly the problem. It is a one-to-one mapping for what the `FactoryBean` does and this is not how we're integrating things in Spring Boot. We usually have an opinion with a sensible default and some easy way to fine tune well-known use cases. For instance, the Spring Batch integration looks for `Job` beans and execute them on startup by default. 

I am not denying that having some kind of scheduling solution in Spring Boot is interesting but I'd rather focus on the 80% use cases first. And this PR doesn't do that.

Maybe those who voted could share their most frequent use case and we could try to derive a proper default behaviour from that?

vpavic commented:
> It is a one-to-one mapping for what the ```FactoryBean``` does and this is not how we're integrating things in Spring Boot. We usually have an opinion with a sensible default and some easy way to fine tune well-known use cases. For instance, the Spring Batch integration looks for ```Job``` beans and execute them on startup by default.

@snicoll I'm confused by your comment. Auto-configuration for ```SchedulerFactoryBean``` implemented in this PR offers automatic registration of any ```JobDetail```, ```Calendar```, ```Trigger``` and various Quartz listener beans with ```SchedulerFactoryBean```. Normally you need to register those manually with ```SchedulerFactoryBean```. Also if there is a ```DataSource``` available persistent ```JobStore``` will be enabled.

snicoll commented:
I guess what I tried to say is that it maps all the features of the `FactoryBean` and maybe it shouldn't do that for a first version (i.e. we should concentrate on the regular use case and let users configure things manually if they use something a bit more specific). We usually fine tune in increment with user's feedback.

vpavic commented:
OK, I'm open to suggestions.

Persistent ```JobStore``` as well as ```JobDetail```s, ```Calendar```s and ```Trigger```s are the bare minimum one would expect. Listeners and ```JobFactory``` might not be used that often and one could live without them being auto-configured I guess.

As far as configuration properties are concerned, we regulary use ```taskExecutor``` to specify executor, ```applicationContextSchedulerContextKey``` to get hold of Spring beans within jobs, ```startupDelay``` and ```waitForJobsToCompleteOnShutdown``` to fine tune startup/shutdown behavior and additional ```quartzProperties``` (to configure clustering and fine-tune ```JobStore``` configuration).

vpavic commented:
I've updated the PR with following changes:

- amount of configuration properties significantly reduced
- auto configuration now picks up only ```JobDetail```, ```Trigger``` and ```Calendar``` beans
- introduced ```QuartzDatabaseInitializer``` with sensible defaults (inspired by ```BatchDatabaseInitializer```)
- introducted ```QuartzSchedulerFactoryBeanCustomizer```
- amount of code reduced by approximately 20% from the original PR
- rebased on to the current master

@snicoll, @philwebb, please review the changes, I'm looking forward to your feedback.

vpavic commented:
Also this PR is another use case which would benefit from #5082.

cemo commented:
:+1:

jgoldhammer commented:
+1

vpavic commented:
Any feedback on the [latest changes](https://github.com/spring-projects/spring-boot/pull/4299#issuecomment-189615197)?
Considering there are many votes on this PR, is there anything else that can be done to have this merged?

mouellet commented:
+1

jgoldhammer commented:
+1

ptahchiev commented:
+1 from me as well. Any updates here?

ptahchiev commented:
It could be also interesting to provide a redis `JobStore` like this project: https://github.com/jlinn/quartz-redis-jobstore

vpavic commented:
@ptahchiev That is an interesting option indeed. Looking at the documentation of the project I guess it should be easily configurable using ```spring.quartz.properties.*``` configuration properties provided by this PR.

bassista commented:
+1

ankurbhakta commented:
+1

steve-oakey commented:
+1

mymasse commented:
+1

vpavic commented:
I've rebased this on to the current master.

Some further notes:

- the implementation does not depend on #5082 as the auto-configuration will make use of `Executor` if it's available, regardless of if it was provided by Boot or user
- #6543 would help improve the database initializer
- quartz-scheduler/quartz#46 would also improve database initializer since Quartz schema scripts would be available on the classpath

ptahchiev commented:
Just out of curiosity - how many votes does it need to be included in spring-boot?

jgoldhammer commented:
At first, thx for your work. Hope to see it integrated soon...
I would love to see this little configuration to enable autowiring beans into quartz Jobs (http://stackoverflow.com/questions/6990767/inject-bean-reference-into-a-quartz-job-in-spring/15211030#15211030).  I don' t like the other approach to fetch other beans via applicationcontext.

vpavic commented:
@jgoldhammer That's an excellent suggestion! I've updated the PR.

jgoldhammer commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/quartz/AutowireCapableBeanJobFactory.java:
```diff
@@ -0,0 +1,49 @@
+	@Override
+	protected Object createJobInstance(TriggerFiredBundle bundle) throws Exception {
+		Object jobInstance = super.createJobInstance(bundle);
+		this.beanFactory.autowireBean(jobInstance);
```
Can you add initializeBean after autowireBean to call postcontructs of injected beans? 
http://stackoverflow.com/questions/6990767/inject-bean-reference-into-a-quartz-job-in-spring/15211030#comment59382309_15211030

Maybe it is good to have a unit test for this? What do you think?

vpavic replied:
> Can you add initializeBean after autowireBean to call postcontructs of injected beans? 
http://stackoverflow.com/questions/6990767/inject-bean-reference-into-a-quartz-job-in-spring/15211030#comment59382309_15211030

OK, will look into it.

> Maybe it is good to have a unit test for this? What do you think?

If you refer to job auto-wiring support, it's already tested, take a look at `QuartzAutoConfigurationTests#withConfiguredJobAndTrigger`.

poznachowski replied:
Not sure if it's a known limitation, but due to the fact that it uses `Object jobInstance = super.createJobInstance(bundle);`  which is implemented with: `bundle.getJobDetail().getJobClass().newInstance();` from `AdaptableJobFactory` constructor injection in Job beans is not possible.

vpavic replied:
> To make it work, could we simply do: Object job = this.beanFactory.createBean(bundle.getJobDetail().getJobClass()); and then apply properties population from SpringBeanJobFactory ? It seems to work.

If we do this we basically drop the standard `SpringBeanJobFactory` behavior since we don't call the `super.createJobInstance(bundle)` anymore, right?

BTW I wonder if `JobFactory` implementation is Boot's business after all. If there are any improvements in that department I guess more natural place would be Spring Core, together with the rest of Quartz Scheduler support. Thoughts on this @snicoll?

snicoll replied:
constructor injection is not something that is supported currently so that seems quite unrelated to this PR. If you want to support construtor injection, then please raise an issue in the Spring Framework issue tracker. The whole quartz support works currently with properties injection and I don't think that's something we could/should fix in boot.

@vpavic that factory is indeed a bit more opinionated. It's fine to put it here for the time being since it's hidden. We can revisit this if we integrate the feature in the framework. I'll discuss that with @jhoeller

gochev commented:
+1

hanusto commented:
+1

vpavic commented:
I've updated the PR to simplify the database initializer using support introduced in #6543.

Additionally, there seems to be some movement on the Quartz side with regard to inclusion of schema scripts in Quartz artifact so hopefully improvements will be possible in that department as well.

asaikali commented:
+1  a lot of apps that I have used make sure of Quartz and setting it up correctly is not trivial, spring boot can really help here. Hope this gets integrated soon.

rossifederico commented:
+1

zhxxcq commented:
+1

sergeysimonov commented:
+1

akkamburov commented:
+1

laguiar commented:
+1

quartz + (proper) cluster exec it's crucial item for my microservice architecture.

javmain commented:
+1

philwebb commented:
No more +1's please :) We're going to try and get this into Boot 2.0

vpavic commented:
Quartz Scheduler `2.3.0` should be [released in mid-April](https://groups.google.com/d/topic/quartz-scheduler-dev/a1mFkW75PKY/discussion).

This release will include Quartz database schema scripts as a part of `quartz-core` artifact (see quartz-scheduler/quartz#46) which will provide room for improvement of Quartz support proposed by this PR. I'll update the PR shortly after `2.3.0` has been released.

ptahchiev commented:
This project: https://github.com/FlavioF/quartz-scheduler-hazelcast-jobstore
provides implementation of a Quartz Scheduler Job Store using Hazelcast distributed Maps and Sets. Might be interesting to autoconfigure it if hazelcast is found in the classpath.

vpavic commented:
PR updated to align with newly released Quartz `2.3.0`.

poznachowski commented:
As stated here: https://github.com/spring-projects/spring-boot/pull/4299/files?diff=unified#r118478148 current solution disallow using constructor injection (and possibly more things - aspects etc.) in Quartz job.
